### PR TITLE
:bug: Fix resources filenames sanity across OS

### LINKED ISSUE #797: [Bug] Windows: export filenames with ':' become NTFS Alternate Data Streams and break transform
## Summary

On Windows, `crane` / `mta-ops` **export** writes resource filenames that include `:` (for example OpenShift RoleBindings named `system:deployers`). NTFS treats `:` as **Alternate Data Stream (ADS)** syntax, so:

1. The visible file is truncated to a **0-byte** name ending at `system` (no `.yaml`).
2. The real YAML is stored in hidden ADS streams (`deployers.yaml`, `image-builders.yaml`, `image-pullers.yaml`).
3. **transform** then fails when it reads the empty main file:

```text
Object 'Kind' is missing in 'null'
```

This does **not** reproduce on Linux/macOS, where `:` is a valid filename character.

## Environment

- OS: Windows (NTFS)
- Binary: `mta-ops_windows_amd64.exe` / crane Windows amd64
- Cluster: OpenShift (any namespace; OCP auto-creates `system:*` RoleBindings)

## Impact

- Any export that includes resources whose **name** contains `:` is corrupted on disk on Windows.
- Most obvious with default OpenShift RoleBindings:
  - `system:deployers`
  - `system:image-builders`
  - `system:image-pullers`
- End-to-end migration / e2e tests on Windows fail at transform (e.g. MTA-830 path).

## Exact root cause

Export builds filenames like:

```text
RoleBinding_authorization.openshift.io_v1_<ns>_system:deployers.yaml
```

On Windows, that path is interpreted as:

| Component | Meaning |
|---|---|
| `..._system` | Main file (`:$DATA`) — **empty (0 bytes)** |
| `deployers.yaml` | ADS stream holding the real YAML |

`dir` without `/r` only shows the empty main file. `dir /r` reveals the streams.

### Prerequisites

- Windows host
- `mta-ops` / crane Windows binary
- `oc` logged into an OpenShift cluster with a context that can create a project

### 1. Create a project (any empty OCP project is enough)

```bat
oc new-project repro-colon-ads
oc get rolebinding -n repro-colon-ads
```

Confirm RoleBinding names contain colons, e.g. `system:deployers`.

### 2. Export with crane / mta-ops

Use a writable directory under the user profile (e.g. Desktop):

```bat
mkdir %USERPROFILE%\Desktop\crane-ads-repro
cd /d %USERPROFILE%\Desktop\crane-ads-repro

mta-ops_windows_amd64.exe export --context <your-context> --namespace repro-colon-ads --export-dir .\export
```

### 3. Observe truncated / empty files

```bat
dir export\resources\repro-colon-ads
```

Expected bad entries (0 bytes, no `.yaml` suffix):

```text
RoleBinding_authorization.openshift.io_v1_repro-colon-ads_system
RoleBinding_rbac.authorization.k8s.io_v1_repro-colon-ads_system
```

### 4. Prove content is in NTFS Alternate Data Streams

```bat
dir /r export\resources\repro-colon-ads\RoleBinding_*_system
```

Example observed output:

```text
         0 RoleBinding_authorization.openshift.io_v1_repro-colon-ads_system
       965 RoleBinding_authorization.openshift.io_v1_repro-colon-ads_system:deployers.yaml:$DATA
       955 RoleBinding_authorization.openshift.io_v1_repro-colon-ads_system:image-builders.yaml:$DATA
       949 RoleBinding_authorization.openshift.io_v1_repro-colon-ads_system:image-pullers.yaml:$DATA
         0 RoleBinding_rbac.authorization.k8s.io_v1_repro-colon-ads_system
       945 RoleBinding_rbac.authorization.k8s.io_v1_repro-colon-ads_system:deployers.yaml:$DATA
       936 RoleBinding_rbac.authorization.k8s.io_v1_repro-colon-ads_system:image-builders.yaml:$DATA
       969 RoleBinding_rbac.authorization.k8s.io_v1_repro-colon-ads_system:image-pullers.yaml:$DATA
```

### 5. Show transform failure

```bat
mta-ops_windows_amd64.exe transform --export-dir .\export --transform-dir .\transform --overwrite
```

Expected error (wording may vary slightly):

```text
Error: stage 10_KubernetesPlugin: failed to load input resources from ...\export:
file "...\RoleBinding_..._system" is not a valid Kubernetes resource:
Object 'Kind' is missing in 'null'
```

## Expected behavior

- Export should write Windows-safe filenames (sanitize reserved characters such as `:`).
- Transform should be able to read the exported resources on Windows the same as on Linux/macOS.

## Suggested fix direction

Sanitize reserved Windows filename characters when generating export paths (at least `:` ; also consider `<>:"/\|?*`), e.g. replace `:` with `_` (or another safe encoding), and use the same encoding consistently for write and read.

Optionally fail fast on Windows if an unsanitized path would be invalid, instead of silently creating ADS.

## Workaround

Manually delete the empty `RoleBinding_*_system` files before transform (this drops those RoleBindings from the pipeline), or run export/transform on Linux/macOS / WSL.

## Cleanup

```bat
oc delete project repro-colon-ads
rmdir /s /q %USERPROFILE%\Desktop\crane-ads-repro
```

### PR DESCRIPTION
**Pull request by aufi**
Windows systems consider `:` as a special charater, crane should avoid using this character in filenames in file sit creates. Adding replacement of `:` and other potentially dangerous chars with `_` in resources filenames and additional hash to avoid collisions.

Fixes: https://github.com/migtools/crane/issues/797

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved exported resource filenames by replacing Windows-reserved characters with underscores.
  * Added support for safely handling resource names containing colons and other reserved characters.
  * Prevented filename collisions when sanitization or truncation produces matching names.
  * Preserved consistent filenames for namespaced and cluster-scoped resources.
  * Ensured exported filenames remain within supported length limits.

### CODE REVIEW

File: internal/file/file_helper.go [hunk 1]
```diff
@@ -169,6 +169,20 @@ func (opts *PathOpts) GetStageOutputDir(stageName string) string {
...
+// Applied to the joined basename so that exported files are valid on all platforms.
+func sanitizeFilename(name string) string {
+	return strings.Map(func(r rune) rune {
+		switch r {
+		case '<', '>', ':', '"', '/', '\\', '|', '?', '*':
+			return '_'
+		default:
+			return r
+		}
+	}, name)
```
[the remark this answers is not in the document]
Comment (aufi) [reply internal/file/file_helper.go:184]: This is very unlikely to happen, but possible, updated using proposed hash addition to filename.

**Comment by aufi** [comment]
@coderabbitai review

File: internal/file/file_helper.go [hunk 2]
```diff
@@ -177,11 +192,17 @@ func GetResourceFilename(obj unstructured.Unstructured) string {
...
 		obj.GetObjectKind().GroupVersionKind().GroupKind().Group,
 		obj.GetObjectKind().GroupVersionKind().Version,
 		namespace,
 		obj.GetName(),
-	}, "_") + ".yaml"
+	}, "_")
+	sanitized := sanitizeFilename(basename)
+	if sanitized != basename {
+		hash := sha256.Sum256([]byte(basename))
+		sanitized = fmt.Sprintf("%s_%x", sanitized, hash[:4])
```
[the remark this answers is not in the document]
Comment (aufi) [reply internal/file/file_helper.go:212]: This would be too crazy, no.

File: internal/file/file_helper.go [hunk 3]
```diff
@@ -177,11 +192,17 @@ func GetResourceFilename(obj unstructured.Unstructured) string {
...
+	}
+	return sanitized + ".yaml"
```
[the remark this answers is not in the document]
Comment (aufi) [reply internal/file/file_helper.go]: Makes sense, updated.

File: cmd/export/discover_test.go [hunk 4]
```diff
@@ -25,144 +26,6 @@ import (
 	kubetesting "k8s.io/client-go/testing"
 )
 
-// ---------- getFilePath ----------
```
Comment (aufi) [inline cmd/export/discover_test.go:28]: The method is now shared from `file` package, where its tests are placed, removing it from export `package`.

**Comment by aufi** [comment]
@coderabbitai review

File: cmd/export/discover_test.go [hunk 5]
```diff
@@ -493,13 +356,13 @@ func TestWriteResources(t *testing.T) {
...
 	// Namespaced resource written to resourceDir.
-	nsPath := filepath.Join(resourceDir, getFilePath(nsObj))
+	nsPath := filepath.Join(resourceDir, file.GetResourceFilename(nsObj))
 	if _, err := os.Stat(nsPath); err != nil {
 		t.Errorf("namespaced resource file not found: %v", err)
 	}
 
 	// Cluster-scoped resource written to clusterDir.
-	clPath := filepath.Join(clusterDir, getFilePath(clObj))
+	clPath := filepath.Join(clusterDir, file.GetResourceFilename(clObj))
```
[the remark this answers is not in the document]
Comment (aufi) [reply cmd/export/discover_test.go:365]: unit test is enough, another windows-specific E2E test will be added later.

File: internal/file/file_helper_test.go [hunk 6]
```diff
@@ -130,6 +132,198 @@ func TestReadFilesNonExistentDir(t *testing.T) {
...
+	path := file.GetResourceFilename(obj)
+	if len(path) > 255 {
+		t.Errorf("path length %d exceeds 255: %q", len(path), path)
+	}
+	if !strings.HasSuffix(path, ".yaml") {
+		t.Errorf("path does not end with .yaml: %q", path)
+	}
+	if len(path) < 22 {
+		t.Errorf("path too short to contain hash: %q", path)
+	}
```
[the remark this answers is not in the document]
Comment (aufi) [reply internal/file/file_helper_test.go:305]: theoretical test edge case, skipping for now

**Review by msajidmansoori12 — APPROVED** [review]
lgtm , thanks for the quick fix Marek !
